src/VkBootstrap.cpp

```diff
@@ -1182,72 +1190,93 @@ detail::Result<PhysicalDevice> PhysicalDeviceSelector::select() const {
+		phys_dev.suitable = is_device_suitable(phys_dev);
+		if (phys_dev.suitable != PhysicalDevice::Suitable::no) {
+			physical_devices.push_back(phys_dev);
+		}
 	}
 
-	PhysicalDeviceDesc selected_device{};
+	// sort the list into fully and partially suitable devices. use stable_partition to maintain relative order
+	const auto partition_index = std::stable_partition(physical_devices.begin(),
+	    physical_devices.end(),
+	    [](auto const& pd) { return pd.suitable == PhysicalDevice::Suitable::yes; });
 
-	if (criteria.use_first_gpu_unconditionally) {
-		selected_device = phys_device_descriptions.at(0);
-	} else {
-		for (const auto& device : phys_device_descriptions) {
-			auto suitable = is_device_suitable(device);
-			if (suitable == Suitable::yes) {
-				selected_device = device;
-				break;
-			} else if (suitable == Suitable::partial) {
-				selected_device = device;
-			}
-		}
+	// Remove the partially suitable elements if they aren't desired
+	if (selection == DeviceSelectionMode::only_return_fully_suitable) {
+		physical_devices.erase(partition_index, physical_devices.end() - 1);
 	}
 
-	if (selected_device.phys_device == VK_NULL_HANDLE) {
+	return physical_devices;
+}
+
+detail::Result<PhysicalDevice> PhysicalDeviceSelector::select(DeviceSelectionMode selection) const {
+	auto const selected_devices = select_impl(selection);
+
+	if (!selected_devices) return detail::Result<PhysicalDevice>{ selected_devices.error() };
+	if (selected_devices.value().size() == 0) {
 		return detail::Result<PhysicalDevice>{ PhysicalDeviceError::no_suitable_device };
 	}
-	PhysicalDevice out_device{};
-	out_device.physical_device = selected_device.phys_device;
-	out_device.surface = instance_info.surface;
-	out_device.features = criteria.required_features;
-	out_device.extended_features_chain = criteria.extended_features_chain;
-	out_device.properties = selected_device.device_properties;
-	out_device.memory_properties = selected_device.mem_properties;
-	out_device.queue_families = selected_device.queue_families;
-	out_device.defer_surface_initialization = criteria.defer_surface_initialization;
-	out_device.instance_version = instance_info.version;
-
-	out_device.extensions_to_enable.insert(out_device.extensions_to_enable.end(),
-	    criteria.required_extensions.begin(),
-	    criteria.required_extensions.end());
-	auto desired_extensions_supported =
-	    detail::check_device_extension_support(out_device.physical_device, criteria.desired_extensions);
-	out_device.extensions_to_enable.insert(out_device.extensions_to_enable.end(),
-	    desired_extensions_supported.begin(),
-	    desired_extensions_supported.end());
-	return out_device;
+
+	return selected_devices.value().at(0);
+}
+
+// Return all devices which are considered suitable - intended for applications which want to let the user pick the physical device
+detail::Result<std::vector<PhysicalDevice>> PhysicalDeviceSelector::get_suitable_devices(
```

Comment:
How about `select`, `select_devices`, and `select_device_names`? (where the first gets the first one, the second gets the list, and the last returns a vector of strings of devices that 'are suitable')

Reply:
But also your suggested change will break applications, since the return type is different. Now its a vector of devices where before it was just a PhysicalDevice.